testsuite/MDAnalysisTests/analysis/test_density.py

```diff
@@ -154,20 +151,126 @@ class Test_density_from_Universe(object):
+        updating = (mode == "dynamic")
+        self.check_DensityAnalysis(
+            universe.select_atoms(self.selections[mode], updating=updating),
+            self.references[mode]['meandensity'],
+            tmpdir=tmpdir
+        )
+
+    def test_sliced(self, universe, tmpdir):
+        self.check_DensityAnalysis(
+            universe.select_atoms(self.selections['static']),
+            self.references['static_sliced']['meandensity'],
+            tmpdir=tmpdir,
+            runargs=dict(start=1, stop=-1, step=2),
+        )
+
+    def test_userdefn_eqbox(self, universe, tmpdir):
+        with warnings.catch_warnings():
+            # Do not need to see UserWarning that box is too small
+            warnings.simplefilter("ignore")
+            self.check_DensityAnalysis(
+                universe.select_atoms(self.selections['static']),
+                self.references['static_defined']['meandensity'],
+                tmpdir=tmpdir,
+                gridcenter=self.gridcenters['static_defined'],
+                xdim=10.0,
+                ydim=10.0,
+                zdim=10.0,
+            )
+
+    def test_userdefn_neqbox(self, universe, tmpdir):
+        self.check_DensityAnalysis(
+            universe.select_atoms(self.selections['static']),
+            self.references['static_defined_unequal']['meandensity'],
+            tmpdir=tmpdir,
+            gridcenter=self.gridcenters['static_defined'],
+            xdim=10.0,
+            ydim=15.0,
+            zdim=20.0,
+        )
+
+    def test_userdefn_boxshape(self, universe):
+        D = density.DensityAnalysis(
+            universe.select_atoms(self.selections['static']),
+            delta=1.0, xdim=8.0, ydim=12.0, zdim=17.0,
+            gridcenter=self.gridcenters['static_defined']).run()
+        assert D.density.grid.shape == (8, 12, 17)
+
+    def test_warn_userdefn_padding(self, universe):
+        regex = ("Box padding \(currently set at 1\.0\) is not used "
+                 "in user defined grids\.")
+        with pytest.warns(UserWarning, match=regex):
+            D = density.DensityAnalysis(
+                universe.select_atoms(self.selections['static']),
+                delta=self.delta, xdim=100.0, ydim=100.0, zdim=100.0, padding=1.0,
+                gridcenter=self.gridcenters['static_defined']).run(step=5)
+
+    def test_warn_userdefn_smallgrid(self, universe):
+        regex = ("Atom selection does not fit grid --- "
+                "you may want to define a larger box")
+        with pytest.warns(UserWarning, match=regex):
+            D = density.density_from_Universe(
+                universe, select=self.selections['static'],
+                delta=self.delta, xdim=1.0, ydim=2.0, zdim=2.0, padding=0.0,
+                gridcenter=self.gridcenters['static_defined'])
+
+    def test_ValueError_userdefn_gridcenter_shape(self, universe):
+        # Test len(gridcenter) != 3
+        with pytest.raises(ValueError):
```

Comment:
Not that we were doing it before, but is it worth adding a match argument here to check that we are picking up the right error?

Reply:
added